src/ReinforcementLearningZoo/src/algorithms/dqns/NFQ.jl

```diff
@@ -0,0 +1,80 @@
+
+# Keyword arguments
+- `approximator::AbstractApproximator` neural network
+- `num_iterations::Integer` number of value iteration iterations in FQI loop (i.e. the outer loop)
+- `epochs` number of epochs to train neural network per iteration
+- `loss_function::F` loss function of the NN
+- `sampler::BatchSampler{SARTS}` data sampler
+- `rng::R` random number generator
+- `γ::Float32` discount rate
+
+# References
+[1] Riedmiller, M. (2005). Neural Fitted Q Iteration – First Experiences with a Data Efficient Neural Reinforcement Learning Method. In: Gama, J., Camacho, R., Brazdil, P.B., Jorge, A.M., Torgo, L. (eds) Machine Learning: ECML 2005. ECML 2005. Lecture Notes in Computer Science(), vol 3720. Springer, Berlin, Heidelberg. https://doi.org/10.1007/11564096_32
+"""
+Base.@kwdef struct NFQ{A, R, F} <: AbstractLearner
+    action_space::AbstractVector
+    approximator::A
+    num_iterations::Integer = 20
+    epochs::Integer = 100
+    loss_function::F = mse
+    rng::R = Random.default_rng()
+    γ::Float32 = 0.9f0
+end
+
+# function NFQ(;
+#     action_space::AbstractVector,
+#     approximator::A,
+#     num_iterations::Integer = 20,
+#     epochs::Integer = 1000,
+#     loss_function::Any = mse,
+#     rng=Random.default_rng(),
+#     γ::Float32 = 0.9f0,
+#     ) where {A}
+#     NFQ(action_space, approximator, num_iterations, epochs, loss_function, rng, γ)
+# end
+
+# Copied from BasicDQN but sure whether it's appropriate
+@functor NFQ (approximator,)
+
+RLCore.forward(L::NFQ, s::AbstractArray) = RLCore.forward(L.approximator, s)
+
+function RLCore.forward(learner::NFQ, env::AbstractEnv)
+    as = action_space(env)
+    return vcat(repeat(state(env), inner=(1, length(as))), transpose(as)) |> x -> send_to_device(device(learner.approximator), x) |> x->RLCore.forward(learner, x) |> send_to_host |> vec 
+end
+
+# Avoid optimisation in the middle of an episode
+function RLBase.optimise!(::NFQ, ::AbstractStage, ::Trajectory) end
+
+function RLBase.optimise!(learner::NFQ, ::PostEpisodeStage, trajectory::Trajectory)
+    Q = learner.approximator
+    γ = learner.γ
+    loss_func = learner.loss_function
+    as = learner.action_space
+    las = length(as)
+    batch = nothing
+    for b in trajectory
+        batch = b
+    end
+    batch === nothing && return
+    
+    (s, a, r, ss) = batch[[:state, :action, :reward, :next_state]]
+    a = Float32.(a)
+    s, a, r, ss = map(x->send_to_device(device(Q), x), (s, a, r, ss))
+    for i = 1:learner.num_iterations
+        # Make an input x samples x |action space| array -- Q --> samples x |action space| -- max --> samples
+        G = r .+ γ .* (cat(repeat(ss, inner=(1, 1, las)), reshape(repeat(as, outer=(1, size(ss, 2))), (1, size(ss, 2), las)), dims=1) |> x -> maximum(RLCore.forward(Q, x), dims=3) |> vec)
+        for e = 1:learner.epochs
+            Flux.train!((x, y) -> loss_func(RLCore.forward(Q, x), y), params(Q.model), [(vcat(s, a), transpose(G))], Q.optimiser)
+        end
```

Comment:
do I understand correctly that after *each* episode, you will update the NN `epochs` times with the same target return, so `num_iterations*epochs` times with the same batch ? Sounds *very* offpolicy. Is there not supposed to be a new batch sampling at each iteration of the outer loop ? I'm not familiar with the algorithm, I'm just asking to be sure, it just looks very overfitting to a single batch.

Reply:
That is not fully correct. Every iteration in the outer loop, the target is updated using the Bellman equation. Then the NN trains `epochs` times on the target. The estimated Q-function is then used to update the target again with the Bellman equation. The idea is much the same as value iteration in dynamic programming. 

By doing this only occasionally on a larger but fixed batch, higher data efficiency can be achieved. So yes, we are (over)fitting on a single batch, but this batch is intended to be large and representative of the actual environment dynamics, whereas with DQN the batches tend to be smaller (at least in the examples I saw in this repo).

More info can be found in Algorithms for Reinforcement Learning by Szepesvari https://sites.ualberta.ca/~szepesva/rlbook.html or the paper by Riedmiller which I cited earlier.

Reply:
Sure. Well, as long as it correctly reproduces the paper it's fine.